Document context on Azure SQL readonly replica (#529)
bitwarden/contributing-docs: docs/architecture/deep-dives/database-replicas.md

> # Read-only database replicas

Reviewer: ```suggestion
# Read-Only Database Replicas
```

The one exception to sentence case with sections is this initial title.

## Context

> Bitwarden Utilizes

Reviewer 2: ```suggestion
Bitwarden utilizes
```

> [Azure SQL Hyperscale](https://learn.microsoft.com/en-us/azure/azure-sql/database/service-tier-hyperscale?view=azuresql)

Reviewer: ❌ Oh, so it's also complaining about this word. Just add this to the custom words file in the repo.

> [Azure SQL Hyperscale](https://learn.microsoft.com/en-us/azure/azure-sql/database/service-tier-hyperscale?view=azuresql)

Reviewer 2: There is a linebreak between these lines in the rendered page.

> Bitwarden utilizes the

Reviewer: ```suggestion
and its
```

Combine this with the line above it. I didn't initially notices the statement floating above this one.

> We decided to utilize the

Reviewer: ```suggestion
Bitwarden utilizes the
```

> [High Availability Replica](https://learn.microsoft.com/en-us/azure/azure-sql/database/service-tier-hyperscale-replicas?view=azuresql#high-availability-replica)

Reviewer: ```suggestion
[high-availability replica](https://learn.microsoft.com/en-us/azure/azure-sql/database/service-tier-hyperscale-replicas?view=azuresql#high-availability-replica)
```

> available as a read replica. This allows Bitwarden to double the max worker limit (along with

Reviewer: ```suggestion
available as a _read_ replica. This allows Bitwarden to double the max worker limit (along with
```

> that was available to use as a read replica. This allowed Bitwarden to double the max worker limit

Reviewer: ```suggestion
available as a read replica. This allows Bitwarden to double the max worker limit
```

> compute, memory, network throughput) without needing to scale up the primary database. The HA

Reviewer: ```suggestion
compute, memory, and network throughput) without needing to scale up the primary database. The HA
```

> (along with compute, memory, network throughput) without needing to scale up the primary database.

Reviewer: ```suggestion
(along with compute, memory, and network throughput) without needing to scale up the primary database.
```

The HA replica replicates the primary instance by mirroring the transaction log records.

## Vision

> More read-only capabilities should be used in conjunction with read-only replicas - Especially after

Reviewer: ```suggestion
More read-only capabilities should be used in conjunction with read-only replicas, especially after
```

Reviewer: Still need this.

stored procedures have been tuned and there are prominent performance bottlenecks relating to
read-only queries.

## Usage

> We currently use the read-only replica in two areas

Reviewer: ```suggestion
We currently use the read-only replica in two areas:
```

Reviewer: And this.

> 1. Dapper Repositories with the ReadOnlyConnectionString

Reviewer: ```suggestion
1. Dapper repositories with the `ReadOnlyConnectionString`
```

Reviewer: And this.

>    [defined](https://github.com/search?q=repo%3Abitwarden%2Fserver%20path%3A%2F%5Esrc%5C%2FInfrastructure%5C.Dapper%5C%2FRepositories%5C%2F%2F%20ReadOnlyConnectionString&type=code).

Reviewer 2: I'm not sure this link is needed. This is a required constructor parameter for the base Dapper repository class, so all repositories should have it defined. And Github's search isn't great, e.g. GroupRepository [has it defined](https://github.com/bitwarden/server/blob/main/src/Infrastructure.Dapper/AdminConsole/Repositories/GroupRepository.cs) but it doesn't show up in that search.

>    [defined](https://github.com/search?q=repo%3Abitwarden%2Fserver%20path%3A%2F%5Esrc%5C%2FInfrastructure%5C.Dapper%5C%2FRepositories%5C%2F%2F%20ReadOnlyConnectionString&type=code)

Reviewer: ```suggestion
   [defined](https://github.com/search?q=repo%3Abitwarden%2Fserver%20path%3A%2F%5Esrc%5C%2FInfrastructure%5C.Dapper%5C%2FRepositories%5C%2F%2F%20ReadOnlyConnectionString&type=code).
```

> 2. Data engineering pipelines that require direct connection to the database

Reviewer: ```suggestion
2. Data engineering pipelines that require direct connection to the database.
```

> ## Gotchas/Considerations

Reviewer: ```suggestion
## Gotchas / Considerations
```

- [CAP Theorem](https://en.wikipedia.org/wiki/CAP_theorem) applies here. Microsoft does not
  guarantee data consistency and
  [propagation latency](https://learn.microsoft.com/en-us/azure/azure-sql/database/read-scale-out?view=azuresql#data-consistency)
  could vary greatly. For business logic that need consistency, read-only replicas are not a viable
  solution.

> - Lack of observability - Currently there is limited visibility into query performance metrics as

Reviewer: ```suggestion
- Lack of observability -- currently there is limited visibility into query performance metrics as
```

  well as resource utilization on the replica itself.
- Long-running queries on read-only replicas can cause

>   [blocking](https://learn.microsoft.com/en-us/azure/azure-sql/database/read-scale-out?view=azuresql#long-running-queries-on-read-only-replicas)

Reviewer: ```suggestion
  [blocking](https://learn.microsoft.com/en-us/azure/azure-sql/database/read-scale-out?view=azuresql#long-running-queries-on-read-only-replicas).
```
